src/util/sighash.rs

```diff
@@ -0,0 +1,836 @@
+
+    /// Cache for taproot v1 inputs
+    taproot_cache: Option<TaprootCache>,
+}
+
+/// Values cached common between segwit and taproot inputs
+pub struct CommonCache {
+    prevouts: sha256::Hash,
+    sequences: sha256::Hash,
+
+    /// in theory, `outputs` could be `Option` since `NONE` and `SINGLE` doesn't need it, but since
+    /// `ALL` is the mostly used variant by large, we don't bother
+    outputs: sha256::Hash,
+}
+
+/// Values cached for segwit inputs, it's equal to [CommonCache] plus another round of `sha256`
+pub struct SegwitCache {
+    prevouts: sha256d::Hash,
+    sequences: sha256d::Hash,
+    outputs: sha256d::Hash,
+}
+
+/// Values cached for taproot inputs
+pub struct TaprootCache {
+    amounts: sha256::Hash,
+    script_pubkeys: sha256::Hash,
+}
+
+/// Contains outputs of previous transactions.
+/// In the case [SigHashType] variant is `ANYONECANPAY`, [Prevouts::One] may be provided
+pub enum Prevouts<'u> {
+    /// `One` variant allows to provide the single Prevout needed. It's useful for example
+    /// when modifier `ANYONECANPAY` is provided, only prevout of the current input is needed.
+    /// The first `usize` argument is the input index this [TxOut] is referring to.
+    One(usize, &'u TxOut),
+    /// When `ANYONECANPAY` is not provided, or the caller is handy giving all prevouts so the same
+    /// variable can be used for multiple inputs.
+    All(&'u [TxOut]),
+}
+
+const LEAF_VERSION_TAPSCRIPT: u8 = 0xc0;
+
+/// Information related to the script path spending
+pub struct ScriptPath<'s> {
+    script: &'s Script,
+    code_separator_pos: u32,
+    leaf_version: u8,
+}
+
+/// Hashtype of an input's signature, encoded in the last byte of the signature
+/// Fixed values so they can be casted as integer types for encoding
+#[derive(PartialEq, Eq, Debug, Copy, Clone)]
+pub enum SigHashType {
+    /// 0x0: Used when not explicitly specified, defaulting to [SigHashType::All]
+    Default = 0x00,
+    /// 0x1: Sign all outputs
+    All = 0x01,
+    /// 0x2: Sign no outputs --- anyone can choose the destination
+    None = 0x02,
+    /// 0x3: Sign the output whose index matches this input's index. If none exists,
+    /// sign the hash `0000000000000000000000000000000000000000000000000000000000000001`.
+    /// (This rule is probably an unintentional C++ism, but it's consensus so we have
+    /// to follow it.)
+    Single = 0x03,
+    /// 0x81: Sign all outputs but only this input
+    AllPlusAnyoneCanPay = 0x81,
+    /// 0x82: Sign no outputs and only this input
+    NonePlusAnyoneCanPay = 0x82,
+    /// 0x83: Sign one output and only this input (see `Single` for what "one output" means)
+    SinglePlusAnyoneCanPay = 0x83,
+
+    /// Reserved for future use
+    Reserved = 0xFF,
+}
+
+/// Possible errors in computing the signature message
+#[derive(Debug, Eq, PartialEq)]
+pub enum Error {
+    /// Should never happen since we are always encoding, thus we are avoiding wrap the IO error
+    IoError,
```

Comment:
I guess because `?` is more reassuring to see than `unwrap()` ?

Apart from jokes, I based the initial code on `bip143::SigHashCache::encode_signing_data_to` where it's used `?` when calling consensus_encode, but then is simply unwrapped in `bip143::SigHashCache::signature_hash` so it's equivalent to call unwrap any time and remove the `IoError` variant as I did in 9930df36a3674b1537f49b1be9392e96bd464462

Reply:
Yeah, I think since it would be a logic error on our side it's best to panic early. Upstream shouldn't have to bother with something that should never happen. Thx for fixing :)